libraries/hotstuff/chain_pacemaker.cpp

```diff
@@ -0,0 +1,308 @@
+            if (elapsed.count() > 10000000) { // 10-second intervals to print the report
+               fc::microseconds total_us = end - csc_first_time; // total testing walltime so far since 1st request seen
+               int64_t total_secs = total_us.count() / 1000000; // never zero if report interval large enough
+               int64_t avgs = csc_total.count() / total_secs;
+               int64_t avgr = csc_total.count() / csc_reqs;
+               // core contention report
+               ilog("HS-CORE: csc_total_us:${tot} csc_elapsed_s:${secs} csc_avg_us_per_s:${avgs} csc_reqs:${reqs} csc_avg_us_per_req:${avgr}", ("tot", csc_total)("secs",total_secs)("avgs", avgs)("reqs", csc_reqs)("avgr", avgr));
+               fc::microseconds req_total_us; // will compute global stats for all request types
+               fc::microseconds req_max_us;
+               int64_t          req_count = 0;
+               auto it = reqs.begin();
+               while (it != reqs.end()) {
+                  const std::string & req_name = it->first;
+                  reqstat &req = it->second;
+                  int64_t avgr = req.total_us.count() / it->second.count;
+                  // per-request-type performance report
+                  ilog("HS-CORE: ${rn}_total_us:${tot} ${rn}_max_us:${max} ${rn}_reqs:${reqs} ${rn}_avg_us_per_req:${avgr}", ("rn",req_name)("tot", req.total_us)("max",req.max_us)("reqs", req.count)("avgr", avgr));
+                  req_total_us += req.total_us;
+                  if (req_max_us < req.max_us) {
+                     req_max_us = req.max_us;
+                  }
+                  req_count += req.count;
+                  ++it;
+               }
+               // combined performance report
+               int64_t req_avgr = req_total_us.count() / req_count;
+               ilog("HS-CORE: total_us:${tot} max_us:${max} reqs:${reqs} avg_us_per_req:${avgr}", ("tot", req_total_us)("max",req_max_us)("reqs", req_count)("avgr", req_avgr));
+               csc_last_report_time = end;
+            }
+         }
+      };
+#else
+      struct csc {  // dummy profiler
+         csc(const string & s) { }
+         void core_in() { }
+         void core_out() { }
+      }
+#endif
+//===============================================================================================
+
+   chain_pacemaker::chain_pacemaker(controller* chain, std::set<account_name> my_producers, bool info_logging, bool error_logging)
+      : _chain(chain),
+        _qc_chain("default"_n, this, std::move(my_producers), info_logging, error_logging)
+   {
+   }
+
+   // Called internally by the chain_pacemaker to decide whether it should do something or not, based on feature activation.
+   // Only methods called by the outside need to call this; methods called by qc_chain only don't need to check for enable().
+   bool chain_pacemaker::enabled() const {
+      return _chain->is_builtin_activated( builtin_protocol_feature_t::instant_finality );
```

Comment:
> `is_builtin_activated` is not thread safe. This can only be safely called from the main thread. Work under #1541

I would have thought that a signal was `emit`ed when a protocol feature is activated? But I don't see it in the code.

Reply:
@greg7mdp even if it was signaled it would only be signaled when first activated. Clients of the emit would have to then have to store it themselves. Also if a fork happens during a protocol feature activation the protocol feature might be rolled back.

Reply:
So there would have to be callbacks for the activation and the rolling back of the feature, so that the state that is duplicated in the chain_pacemaker is always up to date regardless, or is the solution to add a mutex and let the net threads read state written by the main thread, or something else?

Reply:
We believe there is actually no reason to check if the protocol feature is enabled. The code will just not be called or not relevant if not enabled.